src/couch/src/couch_debug.erl

```diff
@@ -620,6 +635,90 @@ info_size(InfoKV) ->
         {binary, BinInfos} -> lists:sum([S || {_, S, _} <- BinInfos]);
         {_, V} -> V
     end.
+resource_hoggers(MemoryInfo, InfoKey) ->
+    KeyFun = fun
+        ({_Pid, _Id, undefined}) -> undefined;
+        ({_Pid, _Id, DataMap}) -> maps:get(InfoKey, DataMap)
+    end,
+    resource_hoggers(MemoryInfo, InfoKey, KeyFun).
+
+resource_hoggers(MemoryInfo, InfoKey, KeyFun) ->
+    HoggersData = resource_hoggers_data(MemoryInfo, InfoKey, KeyFun),
+    TableSpec = [
+        {50, left, name},
+        {19, centre, InfoKey},
+        {id}
+    ],
+    print_table(HoggersData, TableSpec).
+
+resource_hoggers_data(MemoryInfo, InfoKey, KeyFun) when is_atom(InfoKey) ->
+    resource_hoggers_data(MemoryInfo, InfoKey, KeyFun, 20).
+
+resource_hoggers_data(MemoryInfo, InfoKey, KeyFun, N) when is_atom(InfoKey) ->
+    SortedTuples = resource_hoggers_data(MemoryInfo, InfoKey, KeyFun, undefined),
```

Comment:
This would create an infinite loop. I think you need an extra guard on this clause to prevent it `is_integer(N)` (or swap the clauses so `undefined` is handled before integer value).

Reply:
The following doesn't terminates (even if I add extra guard)

```erlang
MI = couch_debug:memory_info(erlang:processes()).
couch_debug:resource_hoggers(MI, memory).
```

Reply:
The infinite loop is here:
```
resource_hoggers(MemoryInfo, InfoKey, KeyFun) ->
    HoggersData = resource_hoggers_data(MemoryInfo, InfoKey, KeyFun),
    TableSpec = [
        {50, left, name},
        {19, centre, InfoKey},
        {id}
    ],
    print_table(HoggersData, TableSpec).
```

Reply:
This is my result with the extra guard:
```
(node1@127.0.0.1)2> couch_debug:resource_hoggers(MI, memory).
|name                                              |      memory       |id
** exception error: no function clause matching
                    couch_debug:'-print_table/2-fun-0-'({1945464}) (src/couch_debug.erl, line 837)
     in function  lists:foreach/2 (lists.erl, line 1342)
     in call from couch_debug:print_table/2 (src/couch_debug.erl, line 836)
```

Reply:
Looks like a misalignment between `print_table/2` and `resource_hoggers_data`.

Reply:
Making my own print function to better fit the data structure returned by `resource_hoggers_data`:
```
(node1@127.0.0.1)4> couch_debug:resource_hoggers(MI, memory).
id | memory
"user[<0.70.0>]" | 2792
"tls_sup[<0.116.0>]" | 5840
"tls_server_sup[<0.118.0>]" | 8936
"tls_server_session_ticket_sup[<0.120.0>]" | 2752
"tls_connection_sup[<0.117.0>]" | 2752
"tls_client_ticket_store[<0.114.0>]" | 2752
"timer_server[<0.155.0>]" | 24784
"supervisor_bridge:user_sup/1[<0.69.0>]" | 13712
"standard_error_sup[<0.67.0>]" | 2792
"standard_error[<0.68.0>]" | 42304
"ssl_upgrade_server_session_cache_sup[<0.122.0>]" | 2752
"ssl_sup[<0.110.0>]" | 5840
"ssl_server_session_cache_sup[<0.121.0>]" | 2752
"ssl_pem_cache[<0.112.0>]" | 2832
"ssl_manager[<0.113.0>]" | 2832
"ssl_listen_tracker_sup[<0.119.0>]" | 2752
"ssl_connection_sup[<0.115.0>]" | 3976
"ssl_admin_sup[<0.111.0>]" | 8896
"socket_registry[<0.7.0>]" | 2624
"smoosh_sup[<0.822.0>]" | 2792
ok
```

Reply:
Looks good now

```
> MI = couch_debug:memory_info(erlang:processes()).
> couch_debug:resource_hoggers(MI, memory).
|                        id                        |       memory
|         couch_prometheus_server[<0.280.0>]       |      1803680
|                 config[<0.137.0>]                |      1115184
|               code_server[<0.51.0>]              |       689528
|             erlang:apply/2[<0.1040.0>]           |       426664
|         couch_stats_aggregator[<0.184.0>]        |       284744
|            couch_log_server[<0.168.0>]           |       266128
|             erlang:apply/2[<0.1039.0>]           |       264576
|  ouch_epi_functions_gen_chttpd_handlers[<0.146.0>|       263960
|     couch_epi_data_gen_flags_config[<0.159.0>]   |       176360
|             erl_prim_loader[<0.10.0>]            |       176288
|           couch_proc_manager[<0.320.0>]          |       142956
|    couch_epi_functions_gen_couch_db[<0.150.0>]   |       109312
|   couch_epi_functions_gen_chttpd_auth[<0.143.0>] |       109312
|  couch_epi_data_gen_dreyfus_black_list[<0.154.0>]|       75864
|         couch_db_updater:init/1[<0.401.0>]       |       68080
|     couch_epi_functions_gen_chttpd[<0.148.0>]    |       68072
|              standard_error[<0.69.0>]            |       67960
|              config_event[<0.138.0>]             |       56608
|  couch_epi_functions_gen_global_changes[<0.161.0>|       55128
|  couch_epi_functions_gen_feature_flags[<0.157.0>]|       55128
ok
```